torch/csrc/jit/ir_import.cpp

```diff
@@ -0,0 +1,559 @@
+#include "torch/csrc/jit/ir_import.h"
+#include "torch/csrc/onnx/onnx.pb.h"
+#include "torch/csrc/jit/ir.h"
+
+#include <ATen/ATen.h>
+
+#include <unordered_map>
+#include <vector>
+#include <string>
+
+#include "third_party/nanopb/pb_decode.h"
+
+namespace torch { namespace jit {
+
+namespace {
+
+// Deserialized data
+
+struct Tensor_ {
+  std::vector<int64_t> dims;
+  std::vector<uint8_t> raw_data;
+  onnx_TensorProto_DataType data_type;
+};
+
+struct AttributeValue_ {
+  std::string name;
+  onnx_AttributeProto_AttributeType type;
+  double f;
+  int64_t i;
+  std::string s;
+  Tensor_ t;
+  std::string g;
+  std::vector<double> fs;
+  std::vector<int64_t> is;
+  std::vector<std::string> ss;
+  std::vector<Tensor_> ts;
+  std::vector<std::string> gs;
+};
+
+struct Value_ {
+  std::string name;
+};
+
+struct Node_ {
+  std::string op_type;
+  std::vector<std::string> inputs;
+  std::vector<std::string> outputs;
+  std::vector<AttributeValue_> attrs;
+};
+
+struct Graph_ {
+  std::vector<Value_> inputs;
+  std::vector<Value_> outputs;
+  std::vector<Node_> nodes;
+  std::vector<Tensor_> initializers;
+};
+
+struct Model_ {
+  Graph_ graph;
+};
+
+
+// Readers
+
+struct ReaderBase {
+  ReaderBase() {}
+  ReaderBase(pb_callback_t& cb) {
+    initialize_callback(cb);
+  }
+
+  void initialize_callback(pb_callback_t& cb) {
+    cb.funcs.decode = ReaderBase::decode;
+    cb.arg = this;
+  }
+
+  virtual void decode(pb_istream_t *stream) = 0;
+
+  static bool decode(pb_istream_t *stream, const pb_field_t *, void **_self) {
+    ReaderBase* self = *reinterpret_cast<ReaderBase* const *>(_self);
+    try {
+      self->decode(stream);
+    }
+    catch (const std::runtime_error& error) {
```

Comment:
Why do we handle this in here? This causes some `Reader`s to silently ignore errors (e.g. you don't check the return value in `Reader<std::vector<T>>`). It would be better to handle the exceptions in the caller if that's what's needed.